File: uploadservice/src/main/java/net/gotev/uploadservice/HttpUploadTaskParameters.java
package net.gotev.uploadservice;

import android.os.Parcel;
import android.os.Parcelable;

import net.gotev.uploadservice.data.NameValue;
import net.gotev.uploadservice.extensions.StringExtensionsKt;

import java.util.ArrayList;

/**
 * Class which contains specific parameters for HTTP uploads.
 *
 * @author gotev (Aleksandar Gotev)
 */
public final class HttpUploadTaskParameters implements Parcelable {

    public String customUserAgent;
    public String method = "POST";
    public boolean usesFixedLengthStreamingMode = true;
    private ArrayList<NameValue> requestHeaders = new ArrayList<>(10);
    private ArrayList<NameValue> requestParameters = new ArrayList<>(10);

    public HttpUploadTaskParameters() {

    }

    // This is used to regenerate the object.
    // All Parcelables must have a CREATOR that implements these two methods
    public static final Parcelable.Creator<HttpUploadTaskParameters> CREATOR =
            new Parcelable.Creator<HttpUploadTaskParameters>() {
                @Override
                public HttpUploadTaskParameters createFromParcel(final Parcel in) {
                    return new HttpUploadTaskParameters(in);
                }

                @Override
                public HttpUploadTaskParameters[] newArray(final int size) {
                    return new HttpUploadTaskParameters[size];
                }
            };

    @Override
    public void writeToParcel(Parcel parcel, int arg1) {
        parcel.writeString(method);
        parcel.writeString(customUserAgent);
        parcel.writeByte((byte) (usesFixedLengthStreamingMode ? 1 : 0));
        parcel.writeList(requestHeaders);
        parcel.writeList(requestParameters);
    }

    private HttpUploadTaskParameters(Parcel in) {
        method = in.readString();
        customUserAgent = in.readString();
        usesFixedLengthStreamingMode = in.readByte() == 1;
        in.readList(requestHeaders, NameValue.class.getClassLoader());
        in.readList(requestParameters, NameValue.class.getClassLoader());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public boolean isCustomUserAgentDefined() {
        return customUserAgent != null && !"".equals(customUserAgent);
    }

    public HttpUploadTaskParameters addHeader(String name, String value) {
        if (!StringExtensionsKt.isASCII(name) || !StringExtensionsKt.isASCII(value))
            throw new IllegalArgumentException("Header " + name + " and its value " + value + " must be ASCII only! Read http://stackoverflow.com/a/4410331");

        requestHeaders.add(new NameValue(name, value));
        return this;
    }

    public ArrayList<NameValue> getRequestHeaders() {
        return requestHeaders;
    }

    public HttpUploadTaskParameters addParameter(String name, String value) {
        requestParameters.add(new NameValue(name, value));
        return this;
    }

    public ArrayList<NameValue> getRequestParameters() {
        return requestParameters;
    }
}

File: uploadservice/src/main/java/net/gotev/uploadservice/data/HttpUploadTaskParameters.kt
package net.gotev.uploadservice.data

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import net.gotev.uploadservice.persistence.Persistable
import net.gotev.uploadservice.persistence.PersistableData
import java.util.ArrayList

/**
 * Class which contains specific parameters for HTTP uploads.
 */
@Parcelize
data class HttpUploadTaskParameters(
    var method: String = "POST",
    var usesFixedLengthStreamingMode: Boolean = true,
    val requestHeaders: ArrayList<NameValue> = ArrayList(5),
    val requestParameters: ArrayList<NameValue> = ArrayList(5)
) : Parcelable, Persistable {

    companion object : Persistable.Creator<HttpUploadTaskParameters> {
        private object CodingKeys {
            const val method = "method"
            const val fixedLength = "fixedLength"
            const val headers = "headers"
            const val parameters = "params"
        }

        private fun List<PersistableData>.toNameValueArrayList() =
            ArrayList(map { NameValue.createFromPersistableData(it) })

        override fun createFromPersistableData(data: PersistableData) = HttpUploadTaskParameters(
            method = data.getString(CodingKeys.method),
            usesFixedLengthStreamingMode = data.getBoolean(CodingKeys.fixedLength),
            requestHeaders = try {
                data.getArrayData(CodingKeys.headers).toNameValueArrayList()
            } catch (exc: Throwable) {
                ArrayList()
            },
            requestParameters = try {
                data.getArrayData(CodingKeys.parameters).toNameValueArrayList()
            } catch (exc: Throwable) {
                ArrayList()
            }
        )
    }

    override fun toPersistableData() = PersistableData().apply {
        putString(CodingKeys.method, method)
        putBoolean(CodingKeys.fixedLength, usesFixedLengthStreamingMode)
        putArrayData(CodingKeys.headers, requestHeaders.map { it.toPersistableData() })
        putArrayData(CodingKeys.parameters, requestParameters.map { it.toPersistableData() })
    }
}
